### PR TITLE
Support self-hosted runner groups

### PR DESCRIPTION
**Pull request by renjithr23**
relates to issue #1594

### CODE REVIEW

File: github/actions_runner_groups.go [hunk 1]
```diff
@@ -0,0 +1,286 @@
+// Copyright 2020 The go-github AUTHORS. All rights reserved.
```
Comment (gmlewis) [inline github/actions_runner_groups.go:1]:
```suggestion
// Copyright 2021 The go-github AUTHORS. All rights reserved.
```

File: github/actions_runner_groups.go [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+	Name                     *string `json:"name,omitempty"`
+	Visibility               *string `json:"visibility,omitempty"`
+	Default                  *bool   `json:"default,omitempty"`
+	SelectedRepositoriesURL  *string `json:"selected_repositories_url,omitempty"`
+	RunnersURL               *string `json:"runners_url,omitempty"`
+	Inherited                *bool   `json:"inherited,omitempty"`
+	AllowsPublicRepositories *bool   `json:"allows_public_repositories,omitempty"`
+}
+
+// RunnerGroups represents a collection of self-hosted runner groups configured for an organization
```
Comment (gmlewis) [inline github/actions_runner_groups.go:25]:
```suggestion
// RunnerGroups represents a collection of self-hosted runner groups configured for an organization.
```

File: github/actions_runner_groups.go [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+type RunnerGroups struct {
+	TotalCount   int            `json:"total_count"`
+	RunnerGroups []*RunnerGroup `json:"runner_groups"`
+}
+
+// CreateRunnerGroupRequest represents a request to create a Runner group for an organization
```
Comment (gmlewis) [inline github/actions_runner_groups.go:31]:
```suggestion
// CreateRunnerGroupRequest represents a request to create a Runner group for an organization.
```

File: github/actions_runner_groups.go [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+type CreateRunnerGroupRequest struct {
+	Name       *string `json:"name,omitempty"`
+	Visibility *string `json:"visibility,omitempty"`
+	// List of repository IDs that can access the runner group.
+	SelectedRepositoryIDs []*int64 `json:"selected_repository_ids,omitempty"`
```
Comment (gmlewis) [inline github/actions_runner_groups.go:36]:
```suggestion
	SelectedRepositoryIDs []int64 `json:"selected_repository_ids,omitempty"`
```

File: github/actions_runner_groups.go [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+	// Runners represent a list of runner IDs to add to the runner group.
+	Runners []*int64 `json:"runners,omitempty"`
```
Comment (gmlewis) [inline github/actions_runner_groups.go:38]:
```suggestion
	Runners []int64 `json:"runners,omitempty"`
```

File: github/actions_runner_groups.go [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+}
+
+// UpdateRunnerGroupRequest represents a request to update a Runner group for an organization
```
Comment (gmlewis) [inline github/actions_runner_groups.go:41]:
```suggestion
// UpdateRunnerGroupRequest represents a request to update a Runner group for an organization.
```

File: github/actions_runner_groups.go [hunk 7]
```diff
@@ -0,0 +1,286 @@
...
+type UpdateRunnerGroupRequest struct {
+	Name       *string `json:"name,omitempty"`
+	Visibility *string `json:"visibility,omitempty"`
+}
+
+// SetRepoAccessRunnerGroupRequest represents a request to replace the list of repositories
+// that can access a self-hosted runner group configured in an organization.
+type SetRepoAccessRunnerGroupRequest struct {
+	// Updated list of repository IDs that should be given access to the runner group.
+	SelectedRepositoryIDs []*int64 `json:"selected_repository_ids"`
```
Comment (gmlewis) [inline github/actions_runner_groups.go:51]:
```suggestion
	SelectedRepositoryIDs []int64 `json:"selected_repository_ids"`
```

File: github/actions_runner_groups.go [hunk 8]
```diff
@@ -0,0 +1,286 @@
...
+}
+
+// SetRunnerGroupRunnersRequest represents a request to replace eplaces the list of
```
Comment (gmlewis) [inline github/actions_runner_groups.go:54]:
```suggestion
// SetRunnerGroupRunnersRequest represents a request to replace the list of
```

File: github/actions_runner_groups.go [hunk 9]
```diff
@@ -0,0 +1,286 @@
...
+// self-hosted runners that are part of an organization runner group.
+type SetRunnerGroupRunnersRequest struct {
+	// Updated list of runner IDs that should be given access to the runner group.
+	Runners []*int64 `json:"runners"`
```
Comment (gmlewis) [inline github/actions_runner_groups.go:58]:
```suggestion
	Runners []int64 `json:"runners"`
```

File: github/actions_runner_groups.go [hunk 10]
```diff
@@ -0,0 +1,286 @@
...
+}
+
+// Lists all self-hosted runner groups configured in an organization.
```
Comment (gmlewis) [inline github/actions_runner_groups.go:61]:
```suggestion
// ListOrganizationRunnerGroups lists all self-hosted runner groups configured in an organization.
```

File: github/actions_runner_groups.go [hunk 11]
```diff
@@ -0,0 +1,286 @@
...
+	groups := &RunnerGroups{}
+	resp, err := s.client.Do(ctx, req, &groups)
+	if err != nil {
+		return nil, resp, err
+	}
+
+	return groups, resp, nil
+}
+
+// Get a specific self-hosted runner group for an organization using its RunnerGroup ID.
```
Comment (gmlewis) [inline github/actions_runner_groups.go:85]:
```suggestion
// GetOrganizationRunnerGroup gets a specific self-hosted runner group for an organization using its RunnerGroup ID.
```

File: github/actions_runner_groups.go [hunk 12]
```diff
@@ -0,0 +1,286 @@
...
+	runnerGroup := new(RunnerGroup)
+	resp, err := s.client.Do(ctx, req, runnerGroup)
+	if err != nil {
+		return nil, resp, err
+	}
+
+	return runnerGroup, resp, nil
+}
+
+// Delete a self-hosted runner group from an organization.
```
Comment (gmlewis) [inline github/actions_runner_groups.go:104]:
```suggestion
// DeleteOrganizationRunnerGroup deletes a self-hosted runner group from an organization.
```

File: github/actions_runner_groups.go [hunk 13]
```diff
@@ -0,0 +1,286 @@
...
+
+	req, err := s.client.NewRequest("DELETE", u, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	return s.client.Do(ctx, req, nil)
+}
+
+// Creates a new self-hosted runner group for an organization.
```
Comment (gmlewis) [inline github/actions_runner_groups.go:118]:
```suggestion
// CreateOrganizationRunnerGroup creates a new self-hosted runner group for an organization.
```

File: github/actions_runner_groups.go [hunk 14]
```diff
@@ -0,0 +1,286 @@
...
+//
+// GitHub API docs: https://docs.github.com/en/rest/reference/actions#create-a-self-hosted-runner-group-for-an-organization
+func (s *ActionsService) CreateOrganizationRunnerGroup(ctx context.Context, org string, createReq *CreateRunnerGroupRequest) (*RunnerGroup, *Response, error) {
```
Comment (gmlewis) [inline github/actions_runner_groups.go:121]:
```suggestion
func (s *ActionsService) CreateOrganizationRunnerGroup(ctx context.Context, org string, createReq CreateRunnerGroupRequest) (*RunnerGroup, *Response, error) {
```

File: github/actions_runner_groups.go [hunk 15]
```diff
@@ -0,0 +1,286 @@
...
+		return nil, resp, err
+	}
+
+	return runnerGroup, resp, nil
+}
+
+// UpdateOrganizationRunnerGroup updates a self-hosted runner group for an organization
+//
+// GitHub API docs: https://docs.github.com/en/rest/reference/actions#update-a-self-hosted-runner-group-for-an-organization
+func (s *ActionsService) UpdateOrganizationRunnerGroup(ctx context.Context, org string, groupID int64, updateReq *UpdateRunnerGroupRequest) (*RunnerGroup, *Response, error) {
```
Comment (gmlewis) [inline github/actions_runner_groups.go:140]:
```suggestion
func (s *ActionsService) UpdateOrganizationRunnerGroup(ctx context.Context, org string, groupID int64, updateReq UpdateRunnerGroupRequest) (*RunnerGroup, *Response, error) {
```

File: github/actions_runner_groups.go [hunk 16]
```diff
@@ -0,0 +1,286 @@
...
+	runnerGroup := new(RunnerGroup)
+	resp, err := s.client.Do(ctx, req, runnerGroup)
+	if err != nil {
+		return nil, resp, err
+	}
+
+	return runnerGroup, resp, nil
+}
+
+// Lists the repositories with access to a self-hosted runner group configured in an organization.
```
Comment (gmlewis) [inline github/actions_runner_groups.go:156]:
```suggestion
// ListRepositoryAccessRunnerGroup lists the repositories with access to a self-hosted runner group configured in an organization.
```

File: github/actions_runner_groups.go [hunk 17]
```diff
@@ -0,0 +1,286 @@
...
+	repos := &ListRepositories{}
+	resp, err := s.client.Do(ctx, req, &repos)
+	if err != nil {
+		return nil, resp, err
+	}
+
+	return repos, resp, nil
+}
+
+// Replaces the list of repositories that have access to a self-hosted runner group configured in an organization.
```
Comment (gmlewis) [inline github/actions_runner_groups.go:175]:
```suggestion
// SetRepositoryAccessRunnerGroup replaces the list of repositories that have access to a self-hosted runner group configured in an organization
```

File: github/actions_runner_groups.go [hunk 18]
```diff
@@ -0,0 +1,286 @@
...
+// with a new List of repositories.
+//
+// GitHub API docs: https://docs.github.com/en/rest/reference/actions#set-repository-access-for-a-self-hosted-runner-group-in-an-organization
+func (s *ActionsService) SetRepositoryAccessRunnerGroup(ctx context.Context, org string, groupID int64, ids *SetRepoAccessRunnerGroupRequest) (*Response, error) {
```
Comment (gmlewis) [inline github/actions_runner_groups.go:179]:
```suggestion
func (s *ActionsService) SetRepositoryAccessRunnerGroup(ctx context.Context, org string, groupID int64, ids SetRepoAccessRunnerGroupRequest) (*Response, error) {
```

File: github/actions_runner_groups.go [hunk 19]
```diff
@@ -0,0 +1,286 @@
...
+
+	req, err := s.client.NewRequest("PUT", u, ids)
+	if err != nil {
+		return nil, err
+	}
+
+	return s.client.Do(ctx, req, nil)
+}
+
+// Adds a repository to the list of selected repositories that can access a self-hosted runner group.
```
Comment (gmlewis) [inline github/actions_runner_groups.go:190]:
```suggestion
// AddRepositoryAccessRunnerGroup adds a repository to the list of selected repositories that can access a self-hosted runner group.
```

File: github/actions_runner_groups.go [hunk 20]
```diff
@@ -0,0 +1,286 @@
...
+// The runner group must have visibility set to selected
```
Comment (gmlewis) [inline github/actions_runner_groups.go:191]:
```suggestion
// The runner group must have visibility set to 'selected'.
```

File: github/actions_runner_groups.go [hunk 21]
```diff
@@ -0,0 +1,286 @@
...
+
+	req, err := s.client.NewRequest("PUT", u, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	return s.client.Do(ctx, req, nil)
+}
+
+// Removes a repository from the list of selected repositories that can access a self-hosted runner group.
```
Comment (gmlewis) [inline github/actions_runner_groups.go:205]:
```suggestion
// RemoveRepositoryAccessRunnerGroup removes a repository from the list of selected repositories that can access a self-hosted runner group.
```

File: github/actions_runner_groups.go [hunk 22]
```diff
@@ -0,0 +1,286 @@
...
+// The runner group must have visibility set to selected
```
Comment (gmlewis) [inline github/actions_runner_groups.go:206]:
```suggestion
// The runner group must have visibility set to 'selected'.
```

File: github/actions_runner_groups.go [hunk 23]
```diff
@@ -0,0 +1,286 @@
...
+
+	req, err := s.client.NewRequest("DELETE", u, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	return s.client.Do(ctx, req, nil)
+}
+
+// Lists self-hosted runners that are in a specific organization group.
```
Comment (gmlewis) [inline github/actions_runner_groups.go:220]:
```suggestion
// ListRunerGroupRunners lists self-hosted runners that are in a specific organization group.
```

File: github/actions_runner_groups.go [hunk 24]
```diff
@@ -0,0 +1,286 @@
...
+	runners := &Runners{}
+	resp, err := s.client.Do(ctx, req, &runners)
+	if err != nil {
+		return nil, resp, err
+	}
+
+	return runners, resp, nil
+}
+
+// Replaces the list of self-hosted runners that are part of an organization runner group
```
Comment (gmlewis) [inline github/actions_runner_groups.go:244]:
```suggestion
// SetRunerGroupRunners replaces the list of self-hosted runners that are part of an organization runner group
```

File: github/actions_runner_groups.go [hunk 25]
```diff
@@ -0,0 +1,286 @@
...
+// with a new list of runners.
+//
+// GitHub API docs: https://docs.github.com/en/rest/reference/actions#set-self-hosted-runners-in-a-group-for-an-organization
+func (s *ActionsService) SetRunerGroupRunners(ctx context.Context, org string, groupID int64, ids *SetRunnerGroupRunnersRequest) (*Response, error) {
```
Comment (gmlewis) [inline github/actions_runner_groups.go:248]:
```suggestion
func (s *ActionsService) SetRunerGroupRunners(ctx context.Context, org string, groupID int64, ids SetRunnerGroupRunnersRequest) (*Response, error) {
```

File: github/actions_runner_groups.go [hunk 26]
```diff
@@ -0,0 +1,286 @@
...
+
+	req, err := s.client.NewRequest("PUT", u, ids)
+	if err != nil {
+		return nil, err
+	}
+
+	return s.client.Do(ctx, req, nil)
+}
+
+// Adds a self-hosted runner to a runner group configured in an organization.
```
Comment (gmlewis) [inline github/actions_runner_groups.go:259]:
```suggestion
// AddRunerGroupRunners adds a self-hosted runner to a runner group configured in an organization.
```

File: github/actions_runner_groups.go [hunk 27]
```diff
@@ -0,0 +1,286 @@
...
+
+	req, err := s.client.NewRequest("PUT", u, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	return s.client.Do(ctx, req, nil)
+}
+
+// Removes a self-hosted runner from a group configured in an organization.
```
Comment (gmlewis) [inline github/actions_runner_groups.go:273]:
```suggestion
// RemoveRunerGroupRunners removes a self-hosted runner from a group configured in an organization.
```

File: github/actions_runner_groups_test.go [hunk 28]
```diff
@@ -0,0 +1,473 @@
+// Copyright 2020 The go-github AUTHORS. All rights reserved.
```
Comment (gmlewis) [inline github/actions_runner_groups_test.go:1]:
```suggestion
// Copyright 2021 The go-github AUTHORS. All rights reserved.
```

**Review by gmlewis — CHANGES_REQUESTED** [review]
Thank you, @renjithr23 !  This is off to a great start.
Please feel free to push as many more commits to this PR as you wish (and please do not use force push) because we will perform a single "Squash and Merge" at the very end when we are ready to merge.

**Comment by renjithr23** [comment]
@gmlewis Thanks for the extensive code review. I made all the changes you've suggested.

File: github/actions_runner_groups.go [hunk 29]
```diff
@@ -0,0 +1,286 @@
...
+		return nil, err
+	}
+
+	return s.client.Do(ctx, req, nil)
+}
+
+// RemoveRunerGroupRunners removes a self-hosted runner from a group configured in an organization.
+// The runner is then returned to the default group.
+//
+// GitHub API docs: https://docs.github.com/en/rest/reference/actions#add-a-self-hosted-runner-to-a-group-for-an-organization
```
Comment (gmlewis) [inline github/actions_runner_groups.go:276]:
```suggestion
// GitHub API docs: https://docs.github.com/en/rest/reference/actions#remove-a-self-hosted-runner-from-a-group-for-an-organization
```

Comment (renjithr23) [reply github/actions_runner_groups.go]: I guess I need to be a little more careful. Thanks again!! @gmlewis . I made the change

Comment (gmlewis) [reply github/actions_runner_groups.go]: > I guess I need to be a little more careful. Thanks again!! @gmlewis . I made the change

No worries!  That's what code reviews are for.  :joy:

**Review by gmlewis — APPROVED** [review]
Thank you, @renjithr23 !
LGTM.

Awaiting second LGTM before merging.

**Review by wesleimp — APPROVED** [review]
LGTM 👌🏼

**Comment by gmlewis** [comment]
Thank you, @wesleimp !
Merging.

**Pull request merged**